asf_heat_pump_suitability/pipeline/suitability/calculate_suitability.py

```diff
@@ -0,0 +1,472 @@
+
+    Returns:
+        pl.DataFrame: max possible score per EPC row
+    """
+    df = df.with_columns(
+        pl.when(pl.col("ruc_two_fold").is_not_null())
+        .then(high_heat_demand_scores.get(tech_type))
+        .otherwise(0)
+        .alias("heat_demand_max"),
+        pl.when(pl.col("OFF GAS").is_not_null())
+        .then(offgas_scores.get(tech_type))
+        .otherwise(0)
+        .alias("off_gas_max"),
+        pl.when(pl.col("Property density (households per KM2)").is_not_null())
+        .then(property_density_scores.get(tech_type))
+        .otherwise(0)
+        .alias("property_density_max"),
+        pl.when(pl.col("msoa_avg_outdoor_space_m2").is_not_null())
+        .then(garden_size_scores.get(tech_type) + external_space_scores.get(tech_type))
+        .otherwise(0)
+        .alias("garden_size_max"),
+        pl.when(pl.col("listed_building").is_not_null())
+        .then(site_regs_scores.get(tech_type))
+        .alias("listed_buildings_max"),
+        pl.when(pl.col("in_conservation_area").is_not_null())
+        .then(site_regs_scores.get(tech_type))
+        .alias("in_cons_area_max"),
+        pl.when(pl.col("property_type").is_not_null())
+        .then(
+            # Note: this only works because not_flat and multiple_props scores are mutually exclusive
+            (not_flat_scores.get(tech_type) + multiple_props_scores.get(tech_type))
+        )
+        .otherwise(0)
+        .alias("property_type_max"),
+        pl.when(pl.col("CURRENT_ENERGY_RATING").is_not_null())
+        .then(epc_threshold_scores.get(tech_type))
+        .otherwise(0)
+        .alias("epc_rating_max"),
+    )
+
+    max_cols = [col for col in df.columns if "max" in col]
+    df = df.with_columns(pl.sum_horizontal(max_cols).alias(f"{tech_type}_max_score"))
+
+    return df.select(["UPRN", f"{tech_type}_max_score"])
+
+
+def filter_df_minimum_features(
+    df: pl.DataFrame, features: list = None, threshold: int = 4
+) -> pl.DataFrame:
+    """
+    Calculate number of non-null features for each row of EPC.
+
+    Args:
+        df: EPC dataset with features.
+        features: list of features to calculate HP suitability
+        threshold: minimum features required to be included
+    """
+    if features is None:
+        features = [
+            "ruc_two_fold",
+            "OFF GAS",
+            "Property density (households per KM2)",
+            "msoa_avg_outdoor_space_m2",
+            "listed_building_grade",
+            "in_conservation_area",
+            "property_type",
+            "CURRENT_ENERGY_RATING",
+        ]
+    df = df.with_columns(
+        (len(features) - pl.sum_horizontal(pl.col(features).is_null()))
+    ).rename(
+        {"literal": "n_features"}
+    )  # Not sure why this is required, naming with alias directly doesnt work fsr
+
+    df = df.filter(pl.col("n_features") >= threshold)
+
+    return df
+
+
```

Comment:
@lizgzil added these 2 new functions below for review and updated the run script below